drivers/counter/counter_mspm0.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2024 Bang & Olufsen A/S, Denmark
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ti/driverlib/dl_timerg.h>
+#include <ti/driverlib/dl_timer.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/counter.h>
+
+LOG_MODULE_REGISTER(counter_mspm0, CONFIG_LOG_DEFAULT_LEVEL);
+
+#define DT_DRV_COMPAT ti_mspm0_counter
+// NOTE: it seems(?) that the clocks support multiple input channels but
+// they must be configured with capture mode
+#define MAX_CHANNELS  1
+
+struct counter_mspm0_channel_data {
+	counter_alarm_callback_t callback;
+	void *user_data;
+};
+
+struct counter_mspm0_config {
+	struct counter_config_info info;
+	GPTIMER_Regs *timer;
+	const DL_Timer_ClockConfig clock_cfg;
+	void (*interrupt_init_function)(const struct device *dev);
+};
+
+struct counter_mspm0_data {
+	const struct counter_mspm0_config *config;
+	struct counter_mspm0_channel_data *channel_data;
+	uint32_t freq;
+	DL_Timer_TimerConfig time_cfg;
+	struct k_sem busy_sem;
+};
+
+static int counter_mspm0_start(const struct device *dev)
+{
+	struct counter_mspm0_data *data = dev->data;
+	if (k_sem_take(&data->busy_sem, K_FOREVER) == 0) {
+		const struct counter_mspm0_config *cfg = dev->config;
+		DL_Timer_reset(cfg->timer);
+		DL_Timer_enablePower(cfg->timer);
+
+		DL_Timer_setClockConfig(cfg->timer, (DL_Timer_ClockConfig *)&cfg->clock_cfg);
+		DL_Timer_initTimerMode(cfg->timer, &data->time_cfg);
+
+		DL_Timer_enableInterrupt(cfg->timer, DL_TIMER_INTERRUPT_ZERO_EVENT);
+		cfg->interrupt_init_function(dev);
+
+		DL_Timer_enableClock(cfg->timer);
+		DL_Timer_startCounter(cfg->timer);
+
+		k_sem_give(&data->busy_sem);
+	}
+	return 0;
+}
+
+static int counter_mspm0_stop(const struct device *dev)
+{
+	struct counter_mspm0_data *data = dev->data;
+	if (k_sem_take(&data->busy_sem, K_FOREVER) == 0) {
+		const struct counter_mspm0_config *cfg = dev->config;
+		DL_Timer_stopCounter(cfg->timer);
+
+		k_sem_give(&data->busy_sem);
+	}
+	return 0;
+}
+
+static int counter_mspm0_set_alarm(const struct device *dev, uint8_t chan,
+				   const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct counter_mspm0_data *data = dev->data;
+	if (k_sem_take(&data->busy_sem, K_FOREVER) == 0) {
+
+		struct counter_mspm0_channel_data *channel_data = &data->channel_data[chan];
+		if (channel_data->callback) {
+			LOG_ERR("Callback is already registered");
+			k_sem_give(&data->busy_sem);
+			return -EBUSY;
+		}
+
+		channel_data->callback = alarm_cfg->callback;
+		channel_data->user_data = alarm_cfg->user_data;
+		data->time_cfg.period = alarm_cfg->ticks;
+
+		k_sem_give(&data->busy_sem);
+	}
+	return 0;
+}
+
+static int counter_mspm0_cancel_alarm(const struct device *dev, uint8_t chan)
+{
+	// NOTE: we only support only 1 channel
```

Comment:
Isn't the entire api built around channels? And now it's implemented so that each channel is a device itself? :wink:

Reply:
That's correct. Not all timers support multiple channels. Plus they also need to be configured in a different mode than the one we are using right now.

Reply:
Let's just keep it as is - hopefully TI will provide a generalized driver at some point. This should be perfect for what we need :1st_place_medal: